CWE-643 XPathInjection on Go

intrigus-lgtm opened the pull request:
An XPath injection query is available for C#, JavaScript, Java but not for Go.
This pull request consist of an XPath Injection query to detect cases in which user input is used unsafely included in an XPath query.

I'm pretty much new to the Go language, so I hope my example code works.
I'm also new to the Go QL flavour so I've added a TODO which I'm currently not sure how to solve.

Query help is inspired/ported from the C# query help.

Code will be autoformatted soon.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.qhelp:
```diff
@@ -0,0 +1,45 @@
+references to include the user input.
+</p>
+<p>
+For exmaple, when using the <code>github.com/moovweb/gokogiri</code> API, this can be done by creating a custom subtype of
```
```suggestion
For example, when using the <code>github.com/moovweb/gokogiri</code> API, this can be done by creating a custom subtype of
```

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,143 @@
+ */
+
+import go
+import semmle.go.dataflow.TaintTracking
```
You shouldn't need this import; it's already imported through `go.qll`.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,143 @@
+class XPathSink extends XPathInjectionSink {
+  XPathSink() {
+    exists(CallExpr call |
+      getAMatchingFunction("xpath", 0, call, this.asExpr()).getName().matches("Compile")
```
A more idiomatic way of writing this would be

```ql
exists(Function f |
  f.hasQualifiedName("https://github.com/antchfx/xpath", "Compile") and
  this = f.getACall().getArgument(0)
)
```

and similar for all the other cases.

intrigus-lgtm replied:
This will not work with wildcards in the name, right?
Should I define something like this and use it instead?
````
  /** Holds if this entity is declared in a package with path `pkg` and matches the given `name`. */
  predicate hasQualifiedName(f, string pkg, string name) {
    pkg = f.getPackage().getPath() and
    f.getName().matches(name)
  }

max-schaefer replied:
Sure it does:

```ql
exists(string name | name.matches("<some pattern>") |
  f.hasQualifiedName("<package>", name)
)
```

intrigus-lgtm replied:
That makes perfect sense.
QL/Logical Programming is still a big paradigm shift for me :)

max-schaefer replied:
Yeah, it can seem a bit magic at times.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.go:
```diff
@@ -0,0 +1,19 @@
+  "github.com/moovweb/gokogiri"
+)
+
+func processRequest(r *http.Request, doc *XmlDocument) {
```
I'm having a hard time getting `gokogiri` to work, but should the import above perhaps be `import "github.com/moovweb/gokogiri/xml"` and this declaration `doc *xml.XmlDocument`?

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.go:
```diff
@@ -0,0 +1,19 @@
+  doc, _ := root.SearchWithVariables("//users/user[login/text()='" + username + "' and password/text() = '" + password + "']/home_dir/text()")
+
+  // GOOD: Uses parameters to avoid including user input directly in XPath expression
+  doc, _ := root.SearchWithVariables("//users/user[login/text()=$username and password/text() = $password]/home_dir/text()")
```
This doesn't look quite right to me. Shouldn't this call have a second argument that provides values for the variables?

max-schaefer replied:
Also, since `doc` has already been declared above this should be `=` instead of `:=`.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.go:
```diff
@@ -0,0 +1,19 @@
+
+  // GOOD: Uses parameters to avoid including user input directly in XPath expression
+  doc, _ := root.SearchWithVariables("//users/user[login/text()=$username and password/text() = $password]/home_dir/text()")
+}
```
We'll need to do something with `doc`, otherwise the compiler gets cross about an unused variable.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.qhelp:
```diff
@@ -0,0 +1,45 @@
+For exmaple, when using the <code>github.com/moovweb/gokogiri</code> API, this can be done by creating a custom subtype of
+<code>xpath.VariableScope</code>, and implementing
+<code>ResolveVariable(string, string)</code> to return the user provided data. This
+custom scope can be specified when calling <code>SearchWithVariables(), EvalXPath() or EvalXPathAsBoolean()</code>.
```
```suggestion
custom scope can be specified when calling <code>SearchWithVariables()</code>, <code>EvalXPath()</code>, or <code>EvalXPathAsBoolean()</code>.
```

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,143 @@
+      or 
+      // TODO: The following cases will have false positives in case a *xpath.Expression is supplied
+      // I don't know how to fix this, since I'm new to the Go QL flavour.
+      // https://github.com/moovweb/gokogiri/blob/a1a828153468a7518b184e698f6265904108d957/xml/node.go#L613
```
You can check the type of the supplied expression. In fact, you could treat any node that isn't a `string` or a `[]byte` as a sanitizer. (And in any case I doubt we'd be tracking taint into an `*xpath.Expression` since we don't have taint models for the functions that construct them.)

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,143 @@
+from DataFlow::PathNode source, DataFlow::PathNode sink, XPathInjectionConfiguration c
+where c.hasFlowPath(source, sink)
+select sink.getNode(), source, sink, "$@ flows to here and is used in an XPath expression.",
+  source.getNode(), "User-provided value"
```
```suggestion
  source.getNode(), "A user-provided value"
```

intrigus-lgtm replied:
This has been copied from https://github.com/Semmle/ql/blob/master/csharp/ql/src/Security%20Features/CWE-643/XPathInjection.ql#L19-L20
So I'm not sure whether I should apply your suggested change.

max-schaefer replied:
Please do; I find the C# message awkward bordering on the ungrammatical.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,143 @@
+
+from DataFlow::PathNode source, DataFlow::PathNode sink, XPathInjectionConfiguration c
+where c.hasFlowPath(source, sink)
+select sink.getNode(), source, sink, "$@ flows to here and is used in an XPath expression.",
```
```suggestion
select sink.getNode(), source, sink, "$@ flows here and is used in an XPath expression.",
```

intrigus-lgtm replied:
Same here.

max-schaefer commented:
^ Pro tip: in the "Files changed" view, you have the option of adding review suggestions to a batch and commit them all at once.

intrigus-lgtm commented:
> ^ Pro tip: in the "Files changed" view, you have the option of adding review suggestions to a batch and commit them all at once.

Thanks for the tip. Even after 5 years of GitHub there is still new stuff to learn :)

Changes:
I autoformatted the query.
I added a sanitizer, that blocks all non string and non []byte nodes.
I removed the gokogiri module in favor of goxpath.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+ *              malicious code by the user.
+ * @kind path-problem
+ * @problem.severity error
+ * @precision high
```
Can you remove the `@precision` for now? We'll fill that in later.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+
+  override predicate isSanitizer(DataFlow::Node node) {
+    not node.asExpr().getType() instanceof StringType or
+    not node.asExpr().getType() instanceof ByteSliceType
```
```suggestion
  exists(Type t | t = node.getType().getUnderlyingType() |
    not t instanceof StringType or not t instanceof ByteSliceType
  )
```

This will take care of named aliases, and might also be slightly more efficient due to the `getType()` being outside `not`.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+class XPathSink extends XPathInjectionSink {
+  XPathSink() {
+    exists(Function f |
+      f.hasQualifiedName("github.com/antchfx/xpath", "Compile%") and
```
You'll need to introduce an auxiliary variable `name` and `.match` it against the pattern like below.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+    )
+    or
+    exists(Function f |
+      f.hasQualifiedName("github.com/antchfx/xpath", "Select%") and
```
Ditto.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+    )
+    or
+    exists(Function f |
+      f.hasQualifiedName("github.com/antchfx/xpath", "MustCompile%") and
```
Ditto.

max-schaefer commented on ql/src/experimental/Security/CWE-643/XPathInjection.ql:
```diff
@@ -0,0 +1,188 @@
+}
+
+from DataFlow::PathNode source, DataFlow::PathNode sink, XPathInjectionConfiguration c, Function f
+where c.hasFlowPath(source, sink) and f.getName().matches("Compile%")
```
I assume `f` is left-over debugging code?
